File: src/main/java/org/academy/api/client/gui/command/PosTexRectDrawCommand.java
package org.academy.api.client.gui.command;

import com.mojang.blaze3d.pipeline.RenderPipeline;
import com.mojang.blaze3d.vertex.VertexConsumer;
import org.academy.api.client.render.TextureBinding;
import org.academy.api.client.render.UniformPayload;
import org.joml.Matrix4f;

import java.util.List;

public class PosTexRectDrawCommand extends DrawCommand {
    protected final float width;
    protected final float height;
    protected final float u0;
    protected final float v0;
    protected final float u1;
    protected final float v1;

    public PosTexRectDrawCommand(
            RenderPipeline pipeline,
            float width,
            float height,
            float u0,
            float v0,
            float u1,
            float v1,
            List<TextureBinding> textures,
            List<UniformPayload<?>> uniforms
    ) {
        super(pipeline, textures, uniforms);
        this.width = width;
        this.height = height;
        this.u0 = u0;
        this.v0 = v0;
        this.u1 = u1;
        this.v1 = v1;
    }

    @Override
    public void generateVertices(VertexConsumer consumer, Matrix4f pose) {
        consumer.addVertex(pose, 0.0F, 0.0F, 0.0F).setUv(u0, v0);
        consumer.addVertex(pose, 0.0F, height, 0.0F).setUv(u0, v1);
        consumer.addVertex(pose, width, height, 0.0F).setUv(u1, v1);
        consumer.addVertex(pose, width, 0.0F, 0.0F).setUv(u1, v0);
    }
}

File: src/main/kotlin/org/academy/api/client/gui/command/PosTexRectDrawCommand.kt
package org.academy.api.client.gui.command

import com.mojang.blaze3d.pipeline.RenderPipeline
import com.mojang.blaze3d.vertex.PoseStack
import org.academy.api.client.gui.render.VertexWriter
import org.academy.api.client.render.TextureBinding
import org.academy.api.client.render.UniformPayload
import org.joml.Vector3f

open class PosTexRectDrawCommand(
    pipeline: RenderPipeline,
    protected val width: Float,
    protected val height: Float,
    protected val u0: Float,
    protected val v0: Float,
    protected val u1: Float,
    protected val v1: Float,
    textures: List<TextureBinding>,
    uniforms: List<UniformPayload<*>>
) : DrawCommand(pipeline, textures, uniforms) {
    override fun generateVertices(writer: VertexWriter, pose: PoseStack.Pose) {
        val matrix = pose.pose()
        val dest = Vector3f()

        writer.beginVertex()
        matrix.transformPosition(0.0f, 0.0f, 0.0f, dest)
        writer.putVec3f(dest.x, dest.y, dest.z)
        writer.putVec2f(u0, v0)

        writer.beginVertex()
        matrix.transformPosition(0.0f, height, 0.0f, dest)
        writer.putVec3f(dest.x, dest.y, dest.z)
        writer.putVec2f(u0, v1)

        writer.beginVertex()
        matrix.transformPosition(width, height, 0.0f, dest)
        writer.putVec3f(dest.x, dest.y, dest.z)
        writer.putVec2f(u1, v1)

        writer.beginVertex()
        matrix.transformPosition(width, 0.0f, 0.0f, dest)
        writer.putVec3f(dest.x, dest.y, dest.z)
        writer.putVec2f(u1, v0)
    }
}
